include/execution.hpp

```diff
@@ -364,52 +380,71 @@ namespace std::execution {
           self.coro_.resume();
         }
       };
+
       template<class R_>
-        class __op : public __op_base {
+        struct __op : __op_base {
+          using promise_type = __promise<R_>;
+          using __op_base::__op_base;
+        };
+
+      template <class R_>
+        struct __promise : __promise_base {
           using R = __t<R_>;
-        public:
-          struct promise_type : __promise_base {
-            template <class A>
-            explicit promise_type(A&, R& r) noexcept
-              : r_(r)
-            {}
 
-            coro::coroutine_handle<> unhandled_done() noexcept {
-              set_done(std::move(r_));
-              // Returning noop_coroutine here causes the __connect_awaitable
-              // coroutine to never resume past the point where it co_await's
-              // the awaitable.
-              return coro::noop_coroutine();
-            }
+          template <class A>
+          explicit __promise(A&, R& r) noexcept
+            : r_(r)
+          {}
 
-            __op get_return_object() noexcept {
-              return __op{
-                coro::coroutine_handle<promise_type>::from_promise(*this)};
-            }
+          coro::coroutine_handle<> unhandled_done() noexcept {
+            set_done(std::move(r_));
+            // Returning noop_coroutine here causes the __connect_awaitable
+            // coroutine to never resume past the point where it co_await's
+            // the awaitable.
+            return coro::noop_coroutine();
+          }
 
-            // Pass through receiver queries
-            template<class... As, invocable<R&, As...> CPO>
-              requires (!__one_of<CPO, set_value_t, set_error_t, set_done_t>)
-            friend auto tag_invoke(CPO cpo, const promise_type& self, As&&... as)
-              noexcept(is_nothrow_invocable_v<CPO, R&, As...>)
-              -> invoke_result_t<CPO, R&, As...> {
-              return ((CPO&&) cpo)(self.r_, (As&&) as...);
-            }
+          __op<R_> get_return_object() noexcept {
+            return __op<R_>{
+              coro::coroutine_handle<__promise>::from_promise(*this)};
+          }
 
-            R& r_;
-          };
+          template <class A>
+          decltype(auto) await_transform(A&& a)
+            noexcept(nothrow_tag_invocable<as_awaitable_t, A, __promise&>){
+            // Dispatch to tag_invoke directly instead of as_awaitable
+            // because we don't want to consider the possibility that
+            // A is a sender and try to convert it to an awaitable.
+            return tag_invoke(as_awaitable, (A&&) a, *this);
```

Comment:
Correct.

Reply:
So why do we need to avoid calling the `as_awaitable()` CPO?

Reply:
Because `as_awaitable()` treats senders as awaitables by default. Here we're trying to use awaitables as senders. Dispatching directly to `tag_invoke` would only pick up end-user customizations and wouldn't try to convert a sender to an awaitable just so we can convert it back to a sender.

Reply:
It saves work.

Reply:
... and this is why: https://github.com/brycelelbach/wg21_p2300_std_execution/pull/292 attn: @lewissbaker